### ISSUE TITLE
C++17: Warning on inheritance from std::iterator [9137]

### ISSUE THREAD

**Issue by alireza-hosseini** [eProsima/Fast-DDS#1275]
Using Fast-DDS in a c++17 windows project raises below warnings about this custom iterator type: https://github.com/eProsima/Fast-DDS/blob/4ec3610181f4bcd0f23c47f2ab0afd6123d22c3e/include/fastdds/rtps/common/LocatorSelector.hpp#L291 

(Compiler: MSVC142)

Warnings:
```
Fast-DDS\include\fastdds/rtps/common/LocatorSelector.hpp(291): warning C4996: 'std::iterator<std::input_iterator_tag,eprosima::fastrtps::rtps::Locator_t,eprosima::fastrtps::rtps::LocatorSelector::IteratorIndex,eprosima::fastrtps::rtps::Locator_t *,eprosima::fastrtps::rtps::Locator_t &>': warning STL4015: The std::iterator class template (used as a base class to provide typedefs) is deprecated in C++17. (The <iterator> header is NOT deprecated.) The C++ Standard has never required user-defined iterators to derive from std::iterator. To fix this warning, stop deriving from std::iterator and start providing publicly accessible typedefs named iterator_category, value_type, difference_type, pointer, and reference. Note that value_type is required to be non-const, even for constant iterators. You can define _SILENCE_CXX17_ITERATOR_BASE_CLASS_DEPRECATION_WARNING or _SILENCE_ALL_CXX17_DEPRECATION_WARNINGS to acknowledge that you have received this warning.
Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.26.28801\include\xutility(5680): note: see declaration of 'std::iterator'
Fast-DDS\include\fastdds/rtps/common/LocatorSelector.hpp(431): warning C4996: 'std::iterator<std::input_iterator_tag,eprosima::fastrtps::rtps::Locator_t,eprosima::fastrtps::rtps::LocatorSelector::IteratorIndex,eprosima::fastrtps::rtps::Locator_t *,eprosima::fastrtps::rtps::Locator_t &>::pointer': warning STL4015: The std::iterator class template (used as a base class to provide typedefs) is deprecated in C++17. (The <iterator> header is NOT deprecated.) The C++ Standard has never required user-defined iterators to derive from std::iterator. To fix this warning, stop deriving from std::iterator and start providing publicly accessible typedefs named iterator_category, value_type, difference_type, pointer, and reference. Note that value_type is required to be non-const, even for constant iterators. You can define _SILENCE_CXX17_ITERATOR_BASE_CLASS_DEPRECATION_WARNING or _SILENCE_ALL_CXX17_DEPRECATION_WARNINGS to acknowledge that you have received this warning.
Fast-DDS\include\fastdds/rtps/common/LocatorSelector.hpp(436): warning C4996: 'std::iterator<std::input_iterator_tag,eprosima::fastrtps::rtps::Locator_t,eprosima::fastrtps::rtps::LocatorSelector::IteratorIndex,eprosima::fastrtps::rtps::Locator_t *,eprosima::fastrtps::rtps::Locator_t &>::reference': warning STL4015: The std::iterator class template (used as a base class to provide typedefs) is deprecated in C++17. (The <iterator> header is NOT deprecated.) The C++ Standard has never required user-defined iterators to derive from std::iterator. To fix this warning, stop deriving from std::iterator and start providing publicly accessible typedefs named iterator_category, value_type, difference_type, pointer, and reference. Note that value_type is required to be non-const, even for constant iterators. You can define _SILENCE_CXX17_ITERATOR_BASE_CLASS_DEPRECATION_WARNING or _SILENCE_ALL_CXX17_DEPRECATION_WARNINGS to acknowledge that you have received this warning.
```

**Comment by MiguelCompany** [comment]
@alireza-hosseini Could you confirm that #1359 solves this issue?

**Comment by alireza-hosseini** [comment]
@MiguelCompany Sorry for the late response, I can't really test this fix since we are using the release version 1.9.3!
Would it be possible for you to add the fix to this specific version?
Thanks.
